draft-ietf-jsonpath-base.md

```diff
@@ -1485,87 +1494,44 @@ Notes:
   expression, and the following rules for function arguments apply to
   its declared result type.
 * Each argument of the function can be used for the declared type of the corresponding declared
   parameter according to one of the following rules:
-   * The argument is a function expression with declared result type that is the same as the declared type of the parameter.
-   * The argument is a literal primitive value and the defined type of the parameter is `ValueType`.
+   * The argument is a function expression with declared result type that is the same as,
+     or a subtype of, the declared type of the parameter.
+   * The argument is a literal primitive value and the declared type of the parameter is `ValueType`.
+   * The argument is a Singular Path and the declared type of the parameter is `SingleNodeType`.
    * The argument is a Singular Path or `filter-path` (which includes
-     Singular Paths), or a function expression with declared result
-     type `NodesType`.  Where the declared type of the parameter is
-     not `NodesType`, a conversion applies.
-
-#### Conversion example
-{:unnumbered}
-
-While functions returning `NodesType` can appear in both comparisons and test expressions,
-path authors must be aware of the conversions and the potential outcomes.
-Specifically, nodelists convert to values and logicals according to {{tbl-typeconv}}:
-
-| Function return        | Converted as Value    | Converted as Logical |
-| :-:                    | :-:                   | :-:                  |
-| empty nodelist         | `Nothing`             | `LogicalFalse`       |
-| single-node nodelist   | the node's JSON value | `LogicalTrue`        |
-| multiple-node nodelist | `Nothing`             | `LogicalTrue`        |
-{: #tbl-typeconv title="Conversion of nodelists to values and logicals"}
-
-For example, given a function `myfunc()` of declared result type
-`NodesType`, the following filter expressions are both valid but can
-have different and unexpected results:
-
-| function result        | `myfunc(@.a)` evaluation (as a test expression)                              | `myfunc(@.a)==42` evaluation (in a comparison)                                |
-| :-:                    | :-                                                                           | :-                                                                            |
-| empty nodelist         | The nodelist is converted to `LogicalFalse`.<br>The expression result is false. | The empty nodelist is converted to `Nothing`.<br>The expression result is false. |
-| single-node nodelist   | The nodelist is converted to `LogicalTrue`.<br>The expression result is true.   | The nodelist is converted to the node's value.<br>The expression result is true. |
-| multiple-node nodelist | The nodelist is converted to `LogicalTrue`.<br>The expression result is true.   | The nodelist is converted to `Nothing`.<br>The expression result is false.       |
```

Comment:
I encourage you to do an exploration of these cases with the system you've defined.  I expect that the table will be the same.  We may still want the table.

Reply:
Thanks @gregsdennis. I agree with the conclusions of your table, but I'm trying to avoid conversion language and make the function side of the spec depend on, _and consistent with_, the semantics already defined by the non-function side of the spec.

Reply:
I removed all of the conversion language from the table.

> ... make the function side of the spec depend on, and consistent with, the semantics already defined by the non-function side of the spec.

So let's replace the headers so that they're not function-centric.

| expression result        | evaluation as a test expression                              | evaluation in a comparison                                |
| :-:                    | :-                                                                           | :-                                                                            |
| empty nodelist         | Interpreted as `LogicalFalse`.<br>The expression result is false. | Interpreted as `Nothing`.<br>The expression result is false. |
| single-node nodelist   | Interpreted as `LogicalTrue`.<br>The expression result is true.   | Interpreted as the node's value.<br>The expression result is true. |
| multiple-node nodelist | Interpreted as `LogicalTrue`.<br>The expression result is true.   | _not well-formed_ |

Reply:
I don't think it's necessary to talk about interpreting an empty nodelist as `LogicalFalse` or a non-empty nodelist as `LogicalTrue`. We can simply say that an empty nodelist evaluated as a test expression yields false and the other cases yeild true.

I'm confused by the phrases "The expression result is [false/true]". I _think_ you mean the comparison result is false/true. Right?

Reply:
Those phrases can be removed if you like, but I think it ties in with the text for filter expressions in general:

> A boolean expression, usually involving the current node, is evaluated and the current node is selected if and only if the expression yields true.

And later, in your edit,

> In the latter case, the test expression tests whether the function result is `LogicalTrue`.

It's assumed from the second statement that there's a distinction between the "true" from the first statement and `LogicalTrue`, but really I think they're the same thing; there's no reason to make a distinction between them.  Maybe that needs to be cleaned up later.

Reply:
This is confused.  The table is about values, not types.

Reply:
The table is about whatever we want it to be about.  Its sole purpose is clarification.

Reply:
My view is that the table is unhelpful. The column headings are misleading. I don't see that the spec needs to talk about _evaluation_ of a test expression. The current language of the spec is that test expressions _yield_ true or false. What this was trying to get across is that test expressions are predicates. Evaluation (literally "turning into a value") feels much more like implementation language.

Similarly, I don't think the various kinds of node lists need to be _evaluated_ in order to specify the outcome of a comparison involving one or more node lists.

Reply:
I think the table can be helpful to readers to clarify things, but we can remove it for now and see how it reads.